test/server/server.test.js

```diff
@@ -0,0 +1,233 @@
+            expect(fetch.data.getSchema.name).toEqual("test");
+            expect(fetch.data.getSchema.schema).toEqual(schema);
+        });
+    });
+
+    describe("Resolvers", () => {
+        it("should not have any resolvers by default", async () => {
+            const fetch = await graphql(Schema, GET_RESOLVERS, root, null, { schemaId: 1, type: "type" });
+            expect(fetch.data.resolvers).toHaveLength(0);
+        });
+
+        it("should throw an error if creating a resolver without schema", async () => {
+            await graphql(Schema, CREATE_DATA_SOURCE, root, null, dataSource);
+
+            let fetch = await graphql(Schema, GET_DATA_SOURCES, root);
+            expect(fetch.data.dataSources).toHaveLength(1);
+
+            fetch = await graphql(Schema, GET_SCHEMAS, root);
+            expect(fetch.data.schemas).toHaveLength(0);
+
+            const create = await graphql(Schema, UPSERT_RESOLVER, root, null, {
+                schemaId: 1, dataSourceId: 1, ...resolver
+            });
+            expect(create.errors).toHaveLength(1);
+        });
+
+        it("should throw an error if creating a resolvers without a data source", async () => {
+            await graphql(Schema, GET_SCHEMA, root, null, { name: "test" });
+
+            let fetch = await graphql(Schema, GET_SCHEMAS, root);
+            expect(fetch.data.schemas).toHaveLength(1);
+
+            fetch = await graphql(Schema, GET_DATA_SOURCES, root);
+            expect(fetch.data.dataSources).toHaveLength(0);
+
+            const create = await graphql(Schema, UPSERT_RESOLVER, root, null, {
+                schemaId: 1, dataSourceId: 1, ...resolver
+            });
+            expect(create.errors).toHaveLength(1);
+        });
+
+        it("should create a resolver and attach it to a data source", async () => {
+            await graphql(Schema, GET_SCHEMA, root, null, { name: "test" });
+            await graphql(Schema, CREATE_DATA_SOURCE, root, null, dataSource);
+
+            await graphql(Schema, UPSERT_RESOLVER, root, null, {
+                schemaId: 1, dataSourceId: 1, ...resolver
+            });
+
+            const fetch = await graphql(Schema, GET_DATA_SOURCE, root, null, { id: 1 });
+            expect(fetch.data.getOneDataSource.resolvers).toHaveLength(1);
+            expect(fetch.data.getOneDataSource.resolvers[0]).toEqual({ id: 1, ...resolver });
+        });
+
+        it("should create a resolver", async () => {
+            await graphql(Schema, GET_SCHEMA, root, null, { name: "test" });
+            await graphql(Schema, CREATE_DATA_SOURCE, root, null, dataSource);
+
+            await graphql(Schema, UPSERT_RESOLVER, root, null, {
+                schemaId: 1, dataSourceId: 1, ...resolver
+            });
+
+            const fetch = await graphql(Schema, GET_RESOLVERS, root, null, {
+                schemaId: 1, type: resolver.type
+            });
+            expect(fetch.data.resolvers).toHaveLength(1);
```

Comment:
Isn't this the same test as the one above except for the final fetch and expectation? Could it be merged into one test?

Reply:
Good point.